File: tlg_bot/kreacher/plugins/speedtest.py
from telethon import events
from kreacher import *
import asyncio
import speedtest


def testspeed(m):
    try:
        test = speedtest.Speedtest()
        test.get_best_server()
        test.download()
        test.upload()
        test.results.share()
        result = test.results.dict()
    except Exception:
        return
    return result


msg_running = """<i>Kreacher is here to serve you.

Running Speedtest...</i> \U0001F4F6"""


@kreacher.on(events.NewMessage(pattern="^/speedtest"))
async def speedtest_function(message):
    m = await message.reply(msg_running, parse_mode="HTML")
    loop = asyncio.get_event_loop()
    result = await loop.run_in_executor(None, testspeed, m)
    output = f"""**Speedtest Results**

**Client:**
**__ISP:__** {result['client']['isp']}
**__Country:__** {result['client']['country']}
  
**Server:**
**__Name:__** {result['server']['name']}
**__Country:__** {result['server']['country']}, {result['server']['cc']}
**__Sponsor:__** {result['server']['sponsor']}
**__Latency:__** {result['server']['latency']} 
**__Ping:__** {result['ping']}"""
    await kreacher.send_file(message.chat.id, result["share"], caption=output)
    await m.delete()

File: kreacher/commands/speedtest.go
package commands

import (
	"fmt"
	"strconv"

	tele "gopkg.in/telebot.v3"
	"santiirepair.dev/kreacher/helpers"
)

func speedtest(c tele.Context) error {
	msg, err := c.Bot().Send(c.Recipient(), stStarted, tele.ParseMode(tele.ModeHTML))

	if err != nil {
		return err
	}

	storedMessage := tele.StoredMessage{ChatID: msg.Chat.ID, MessageID: strconv.Itoa(msg.ID)}
	defer c.Bot().Delete(&storedMessage)

	st, err := helpers.Speedtest()
	if err != nil {
		return c.Send("*_A problem occurred running the speedtest_*", tele.ParseMode(tele.ModeMarkdownV2))
	}

	if st.Share != "" {
		caption := fmt.Sprintf(
			stResult,
			st.Client.ISP,
			st.Client.Country,
			st.Server.Name,
			st.Server.Country,
			st.Server.Sponsor,
			st.Server.Latency,
			st.Ping,
		)

		photo := &tele.Photo{
			File:    tele.FromURL(st.Share),
			Caption: caption,
		}

		return c.Send(photo, tele.ParseMode(tele.ModeHTML))
	}

	return c.Send(stErr, tele.ParseMode(tele.ModeHTML))
}

var stStarted = `
<b><i>Kreacher is here to serve you.</i></b>

<b><i>Running Speedtest...</i></b> 📶
`

var stResult = `
<b>Speedtest Result</b>

<b>Client:</b>
<b><i>ISP:</i></b> %s
<b><i>Country:</i></b> %s

<b>Server:</b>
<b><i>Name:</i></b> %s
<b><i>Country:</i></b> %s
<b><i>Sponsor:</i></b> %s
<b><i>Latency:</i></b> %.2f
<b><i>Ping:</i></b> %.2f
`

var stErr = "<b><i>The speedtest could not be performed</i></b>"
